pkg/codegen/python/gen.go

```diff
@@ -513,8 +514,51 @@ def get_version():
      return _version_str
 `)
 	}
+	if err != nil {
+		return nil, err
+	}
+
+	if pkg.Parameterization != nil {
+		// If a parameterized package is being generated then we _need_ to use package references
+		param := base64.StdEncoding.EncodeToString(pkg.Parameterization.Parameter)
+
+		_, err = fmt.Fprintf(buffer, `
+_package_lock = asyncio.Lock()
+_package_ref = ...
+async def get_package():
+	global _package_ref
+	if _package_ref is ...:
+		async with _package_lock:
+			if _package_ref is ...:
+				monitor = pulumi.runtime.settings.get_monitor()
+				parameterization = resource_pb2.Parameterization(
+					name=%q,
+					version=get_version(),
+					value=base64.b64decode(%q),
+				)
+				try:
+					registerPackageResponse = monitor.RegisterPackage(
+						resource_pb2.RegisterPackageRequest(
+							name=%q,
+							version=%q,
+							download_url=get_plugin_download_url(),
+							parameterization=parameterization,
+						))
+					_package_ref = registerPackageResponse.ref
+				except Exception as e:
+					raise e
+					_package_ref = None
+	if _package_ref is None:
+		raise Exception("The Pulumi CLI does not support package references. Please update the Pulumi CLI.")
+	return _package_ref
```

Comment:
Not sure if we'll want to use `_package_ref` elsewhere, but is it possible (and if so, is it better) to do this as e.g. an immediately executed closure (well, I guess in Python this has to be a decorator, which potentially is not great)? Not sure if this has implications for the definition of the function or not, but just wondering if we can avoid the top-level globals. Fair enough if it's not possible!

Reply:
I think immediately executed closures might be racy with automation api and setting up the monitor. Thus why I leant towards this locked global. It will only execute once, but won't run till a monitor is definitely set up and we need a package ref.